### DISCUSSION TITLE
Can I run the hint kitten on an arbitrary window?

### DISCUSSION THREAD

**Discussion by groves** [kovidgoyal/kitty#5121] [Q&A] [answered]
I have a kitten that executes a ripgrep query for the text currently selected by the mouse by [finding an open shell window and running the command there](https://github.com/groves/catherd/blob/6a687f82d9c8f9286a19e04fff4768de6a8b135d/rg.py#L62). I have a [custom processor for hints](https://github.com/groves/catherd/blob/main/hint_rg.py#L22) that parses the ripgrep results and opens a selected hint in a window open to my editor.

I trigger both the ripgrep and hint kittens with keypresses mapped in kitty.conf. Since the ripgrep kitten doesn't have a UI, I can trigger it while my editor window is active, the kitten can find the shell window, and the command executes in that second window. With the hint kitten, my custom processor only runs after the hinting has started. I have to move to the ripgrep result window before triggering the hint keypress to have it target the ripgrep window. It'd be nice to do all of that without leaving my editor window.

Is there a way I can run some Python code to select the window for hinting? Or is there another way to select the window in the hint mapping? I'd love to share the shell window finding code in between the kitten and the custom processor.

**Comment by kovidgoyal** [comment]
I havent looked at your code but your description sounds rather
convoluted. I *think* what you are trying to do is select some text,
search for it among files in the filesystem and then open the selected
file in an editor? 

The way I would architect that, is map a shortcut to run a script in a
new kitty window passing the current selection to it. Enabel remote
control on that window in the mapping. The script would run ripgrep and
use something like fzf or the choose kitten to pick a result. Once it
has the selected result it can use remote control to open it in the
editor.

**Reply by groves** [reply]
Your interpretation is almost there. I want to select some text, search for it, and then step through multiple of the search results in my editor. It's that last bit that's different than your summary and pushes me to use an existing shell window and not a one-time overlay.

Because I want to keep the search results visible and select another one later, I was hoping to be able to run hinting again whenever I wanted to select the next result. I can't see how to start picking a next result with a single keypress without being able to trigger hinting in another window.

I can make a tool specific to this, but I've enjoyed composing tools with kitty to do this sort of thing thus far. Having kitty manipulate small tools means I can use them independently when I want to do something slightly different. Having a task specific tool makes it less composable.

**Comment by groves** [comment, marked as answer]
I ended up writing a kitten that finds the window I want and calls `boss._run_kitten` to run the hint kitten in that window: https://github.com/groves/catherd/blob/main/hint_shell.py#L24

That works, but reaching into Boss' internal methods doesn't feel great. Is there a better method to call?

I think I could also accomplish what I want more cleanly if the [`--match` option](https://sw.kovidgoyal.net/kitty/remote-control/#cmdoption-kitty-kitten-match) in kitten remote control took a custom Python module to call. Would it be worth my putting together a PR to add that, or is it a bad idea?

**Reply by kovidgoyal** [reply]
I renamed the method so it is no longer an internal method. It has been unchanged for a long time anyway.

**Reply by groves** [reply]
Thanks!

**Comment by kovidgoyal** [comment]
On Tue, May 24, 2022 at 02:51:42PM -0700, Charlie Groves wrote:
> I ended up writing a kitten that finds the window I want and calls `boss._run_kitten` to run the hint kitten in that window: https://github.com/groves/catherd/blob/main/hint_shell.py#L24
> 
> That works, but reaching into Boss' internal methods doesn't feel great. Is there a better method to call?

_run_kitten is the only one that allows you to override the window the
kitten is run in. I've renamed it to run_kitten_with_metadata, since it
is very stable API at this point anyway.

> 
> I think I could also accomplish what I want more cleanly if the [`--match` option](https://sw.kovidgoyal.net/kitty/remote-control/#cmdoption-kitty-kitten-match) in kitten remote control took a custom Python module to call. Would it be worth my putting together a PR to add that, or is it a bad idea?

Seems like unnecessary complication, see above.
